### PR TITLE
*: Refine include of TiDB.h to speedup re-compiling

### PR DESCRIPTION
**Pull request by JaySon-Huang**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tiflash/issues/6233

Problem Summary:

Changing TiDB.h will make about 600 files being re-compile, which cost too much time.

### What is changed and how it works?

```commit-message
*: Refine include of TiDB.h to speedup re-compiling

  * Split `TiDB_fwd.h`, `TiDBTypes.h`
  * Include `TiDB.h`, `TypeMapping.h` in .cpp files instead of .h files as possible to reduce the re-compiling involve files

```

After these changes, the re-compile files down from 600 to 300

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by ngaut** [comment]
Good job.

File: dbms/src/Debug/dbgKVStore/dbgFuncInvestigator.cpp [hunk 1]
```diff
@@ -143,16 +143,16 @@ void dbgFuncFindKey(Context & context, const ASTs & args, DBGInvoker::Printer ou
...
-        size_t handle_column_size = table_info.is_common_handle ? table_info.getPrimaryIndexInfo().idx_cols.size() : 1;
         auto key_size = arg_size / 2;
         std::vector<Field> start_field;
         start_field.reserve(key_size);
         std::vector<Field> end_field;
         end_field.reserve(key_size);
 
-        for (size_t i = 0; i < handle_column_size; i++)
+        const auto & pk_index = table_info.getPrimaryIndexInfo();
+        for (size_t i = 0; i < pk_index.idx_cols.size(); i++)
```
Comment (SeaRise) [inline dbms/src/Debug/dbgKVStore/dbgFuncInvestigator.cpp:153]:
```suggestion
        const auto & pk_index = table_info.getPrimaryIndexInfo();
        for (const auto & idx_col : pk_index.idx_cols)
        {
            auto & column_info = table_info.columns[idx_col.offset];
        }
```

File: dbms/src/Debug/dbgKVStore/dbgFuncInvestigator.cpp [hunk 2]
```diff
@@ -143,16 +143,16 @@ void dbgFuncFindKey(Context & context, const ASTs & args, DBGInvoker::Printer ou
...
         }
-        size_t handle_column_size = table_info.is_common_handle ? table_info.getPrimaryIndexInfo().idx_cols.size() : 1;
         auto key_size = arg_size / 2;
         std::vector<Field> start_field;
         start_field.reserve(key_size);
         std::vector<Field> end_field;
         end_field.reserve(key_size);
 
-        for (size_t i = 0; i < handle_column_size; i++)
+        const auto & pk_index = table_info.getPrimaryIndexInfo();
```
Comment (SeaRise) [inline dbms/src/Debug/dbgKVStore/dbgFuncInvestigator.cpp:152]: Is it possible that `is_common_handle` is false here?

Comment (JaySon-Huang) [reply dbms/src/Debug/dbgKVStore/dbgFuncInvestigator.cpp:152]: No, line 134 ensure `is_common_handle == true` here

File: dbms/src/Debug/dbgKVStore/dbgFuncInvestigator.cpp (hunk 1 above)
Comment (JaySon-Huang) [reply dbms/src/Debug/dbgKVStore/dbgFuncInvestigator.cpp:153]: We can not do that because line 157 require the `i` to get the offset of input `args`

**Review by Lloyd-Pottiger — APPROVED** [review]

**Pull request merged**